ygen/schemaparse.go

```diff
@@ -17,9 +17,79 @@ package ygen
 import (
 	"bytes"
 	"compress/gzip"
+	"encoding/json"
 	"fmt"
+	"strings"
+
+	"github.com/openconfig/goyang/pkg/yang"
 )
 
+// buildJSON tree takes an input set of modules expressed as a slice of yang.Entry
+// structs, the names of directories within the generated code, and a yang.Entry
+// which corresponds to the fake root and returns a JSON serialised tree of the
+// schema for the set of modules. The JSON document that is returned is always
+// rooted on a yang.Entry which corresponds to the root item, and stores all
+// root-level enties (and their subtrees) within the input module set. All
+// YANG directories are annotated in the output JSON with the name of the type
+// they correspond to in the generated code, and the absolute schema path that
+// the entry corresponds to.
+func buildJSONTree(ms []*yang.Entry, dn map[string]string, fakeroot *yang.Entry) ([]byte, error) {
+	rootEntry := &yang.Entry{
+		Dir:        map[string]*yang.Entry{},
+		Annotation: map[string]interface{}{},
+	}
+	for _, m := range ms {
+		annotateChildren(m, dn)
+		for _, ch := range children(m) {
+			if _, ex := rootEntry.Dir[ch.Name]; ex {
+				return nil, fmt.Errorf("overlapping root children for key %s", ch.Name)
+			}
+			rootEntry.Dir[ch.Name] = ch
+		}
+	}
+
+	if fakeroot != nil {
+		rootEntry.Name = fakeroot.Name
+		rootEntry.Annotation["schemapath"] = "/"
+		rootEntry.Annotation["structname"] = dn[fakeroot.Path()]
+		rootEntry.Kind = yang.DirectoryEntry
+	}
+	// Always annotate the root as a fake root, so that it is not treated
+	// as a path element in ytypes.
+	rootEntry.Annotation["isFakeRoot"] = true
```

Comment:
I wonder if we should just always have a fakeroot and assume that in ytypes.

Reply:
Agree. I suggest that we make this change in the next CL.